lib/algos/list/delimiterNonCurrent.js

```diff
@@ -0,0 +1,204 @@
+     * the "stale date" is older than the "beforeDate".
+     * The in-memory "stale date" is then updated with the version's last-modified date to be used for
+     * the following version.
+     * The process stops and returns the available results if either:
+     * - no more metadata key is left to be processed
+     * - the listing reaches the maximum number of key to be returned
+     * - the internal timeout is reached
+     *  @param {String} keyVersionSuffix   - The key to add
+     *  @param {String} value - The value of the key
+     *  @return {number}      - indicates if iteration should continue
+     */
+    addContents(keyVersionSuffix, value) {
+        if (this._reachedMaxKeys()) {
+            return FILTER_END;
+        }
+
+        if (this.start && Date.now() - this.start > DELIMITER_TIMEOUT_MS) {
+            this.IsTruncated = true;
+            this.logger.info('listing stopped after expected internal timeout',
+                {
+                    timeoutMs: DELIMITER_TIMEOUT_MS,
+                    evaluatedKeys: this.evaluatedKeys,
+                });
+            return FILTER_END;
+        }
+        ++this.evaluatedKeys;
+
+        const versionIdIndex = keyVersionSuffix.indexOf(VID_SEP);
+        const key = keyVersionSuffix.slice(0, versionIdIndex);
+        const versionId = keyVersionSuffix.slice(versionIdIndex + 1);
+
+        this.NextKeyMarker = key;
+        this.NextVersionIdMarker = versionId;
+
+        // For a given key, the latest version is skipped since it represents either:
+        // - the current version or
+        // - the last version of the previous truncated listing
+        const isLatestVersion = key !== this.keyName;
+
+        if (isLatestVersion) {
+            this.keyName = key;
+            // The current last-modified date is kept in memory and used as a "stale date" for the following version.
+            this.staleDate = this.getLastModified(value);
+            return FILTER_ACCEPT;
+        }
+
+        // The following version is pushed only if the "stale date" (picked up from the previous version)
+        // is available (JSON.parse has not failed) and, if the "beforeDate" argument is specified,
+        // the "stale date" is older than the "beforeDate".
+        if (this.staleDate && (!this.beforeDate || this.staleDate < this.beforeDate)) {
+            this.Contents.push({ key, value: this.trimMetadataAddStaleDate(value, this.staleDate) });
+            ++this.keys;
+        }
+
+        // The in-memory "stale date" is then updated with the version's last-modified date to be used for
+        // the following version.
+        this.staleDate = this.getLastModified(value);
+
+        return FILTER_ACCEPT;
+    }
+
+    trimMetadataAddStaleDate(value, staleDate) {
+        let ret = undefined;
+        try {
+            ret = JSON.parse(value);
+            ret.staleDate = staleDate;
+            if (value.length >= TRIM_METADATA_MIN_BLOB_SIZE) {
```

Comment:
Today client does not do anything if the `staleDate` is missing.

I am adding a check so that the listing only returns keys with `staleDate`.

The logic I had in mind is that we make our best effort to retrieve the `staleDate`, if parsing fails, the key (with missing staleDate) is not returned and the next listing will likely return it.